File: ruby/triangle/_iterations/1/triangle.rb
class Triangle
  attr_reader :sides

  def initialize(sides)
    @sides = sides
  end

  def valid
    sides.all? { |side| side > 0 && sides.sum >= side * 2 }
  end

  def equilateral?
    valid && sides.uniq.size == 1
  end

  def isosceles?
    valid && sides.uniq.size <= 2
  end

  def scalene?
    valid && sides.uniq.size == 3
  end
end

File: clojure/triangle/_iterations/1/src/triangle.clj
(ns triangle)

(defn- valid? [[a b c :as sides]]
  (and
    (every? pos? sides)
    (<= c (+ a b))))

(defn- is? [a b c fn]
  (let [sides (sort [a b c])]
    (and (valid? sides) (fn sides))))

(defn equilateral? [a b c]
  (is? a b c (partial apply =)))

(defn isosceles? [a b c]
  (is? a b c (fn [[x y z]] (or (= x y) (= y z)))))

(defn scalene? [a b c]
  (is? a b c (fn [[x y z]] (and (not= x y) (not= y z)))))
